### PR TITLE
Dynamic Model schema validation

### LINKED ISSUE #1178: RFC : Elide Dynamic Model Configuration (Tables Only)
## Overview

Elide will support a configuration file format that allows a simplified, text representation of Elide data models.  This representation can be used together with, or in place of standard Java classes to define Elide models.

On startup, the Elide service will generate and compile POJOs using the model definitions from the configuration files.   Elide will support class loading from both the standard classpath as well as the classes compiled at service start.

## Location of Models

Elide will support a single URL configuration setting, _modelConfig_ that can point to a web or filesystem root directory containing the configuration.  Configuration will be stored in a directory, _models_.  There will be separate subdirectory for Elide entities (CRUD models) and for Elide tables (Analytic models).

One or more models can be stored in a single configuration file.  Optionally, the model files (tarball archive) can be signed with PGP to verify their integritity (See security).

All model files support variable substitution preprocessing by referencing variables from a variables.hjson file.

```
models
└───entities
    │   entity1.hjson  
└───tables
    │   table1.hjson
    │   table2.hjson
└───variables.hjson
└───security.hjson
└───models.asc
```

## Model Format

Model files are written in [HJSON](https://hjson.github.io/) - a human friendly configuration format that resembles relaxed JSON & allows for comments.  Each model file can contain a single model or an array of models.

### Example

```
{
  name: PlayerStats
  table: player_stats
  schema: gamedb
  description:
  '''
  A long description
  '''
  cardinality : large
  readAccess : A user is admin or is a player in the game

  joins: [
      {  
         name: playerCountry
         to: country
         type: toOne
         definition: '${to}.id = ${from}.country_id'
      }
  ]

  measures : [
     {  
        name : highScore
        type : number
        definition: 'MAX(score)'
     }
  ]

  dimensions : [
     { 
       name : countryCode
       type : text
       definition : playerCountry.isoCode
     },
     { 
       name : createdOn
       type : time 
       definition : create_on
       grains: [{{hour}}, {{day}}]
     }
  ]
}
```

### Table Properties

Table models have the following properties:

| Property                      | Required   | Description                               | Default Value |
| ----------------------------- | ---------- | ----------------------------------------- | ------------- |
| name                          | Yes        | The name of the model.  This will be the same as the POJO class name. | N/A |
| schema                        | No         | The database or schema where the model lives.  |  empty |
| table                         | One of (table, sql, or extends)        | The physical table name where the model lives.  |  empty |
| sql                           | One of (table, sql, or extends)           | SQL query which is used to populate the table.  |  empty |
| extends                   | One of (table, sql, or extends)          | Extends another logical table. |  empty |
| hidden                     | No          | Whether this table is exposed via API metadata | false |
| description                   | No         | A long description of the model.          | The model name |
| cardinality                   | No         | The number of rows in the table: (tiny, small, medium, large, huge).  The relative sizes are decided by the table designer(s). | tiny |
| readAccess                    | No         | Read permission for the table.            | Allow All     |
| joins                         | No         | Describes SQL joins to other tables for column references. | empty |
| measures                      | No         | Zero or more metric definitions.          | empty |
| dimensions                    | Yes        | One or more dimension definitions.        | empty |
| tags                                | No        | An array of string based tags       | empty |

### Joins

Joins describe the SQL expression necessary to join two physical tables.  Joins can be used when defining dimension columns that reference other tables.

| Property                      | Required   | Description                               | Default Value |
| ----------------------------- | ---------- | ----------------------------------------- | ------------- |
| name                          | Yes        | The name of the relationship.             | N/A           |
| to                            | Yes        | The name of the table that is being joined to | N/A |
| type                          | Yes        | Either toOne or toMany                    | N/A           |
| definition                    | Yes        | Templated SQL expression that represents the ON clause of the join | N/A |

Join expressions are designed to join two tables together (not more).  Columns in the join expression are defined with a simplate template syntax:

`${to|from}.columnName`

`${from}` contains the name of the current table/model referenced in the join.  `${to}` contains the name of the table that is being joined to the current table.   The `columnName` refers to either a either a model attribute or a physical SQL column name (with the former taking precedence over the latter).

### Measures

Measures represent metrics that can be aggregated at query time.  They have the following properties:

| Property                      | Required   | Description                               | Default Value |
| ----------------------------- | ---------- | ----------------------------------------- | ------------- |
| name                          | Yes        | The name of the metric.  This will be the same as the POJO field name. | N/A |
| description                   | No         | A long description of the metric.          | The metric name |
| category                       | No         | Category for grouping                                  | N/A            |
| hidden                     | No          | Whether this metric is exposed via API metadata | false |
| readAccess                    | No         | Read permission for the metric.            | Allow All     |
| definition                    | Yes        | The definition of the metric (See below)   | N/A           |
| type                          | No         | The data type of the field (See types)     | number          |

Metric definitions are extensible objects that contain a _type_ field and one or more additional attributes.  Each type is tied to logic in Elide that generates a metric function.  These functions can optionally take arguments and are responsible for generating templated SQL at query time.

Most metric definitions however take no arguments and can be represented with simple SQL expressions.  To simplify the common case, _definition_ can be mapped to a simple templated string instead of an object:

`definition: 'SUM(column)'`

The SQL templates can reference columns:
1. by name: `column`
2. by a dot separated path to a column: `player.session.score`.

When using dot separated paths, all segments (excluding the final segment) represent _join_ names originating from the current table.  The final segment represents the column in the final joined table.

### Dimensions

Dimensions represent labels for measures.  Dimensions are used to filter and group measures.  They have the following properties:

| Property                      | Required   | Description                               | Default Value |
| ----------------------------- | ---------- | ----------------------------------------- | ------------- |
| name                          | Yes        | The name of the dimension.  This will be the same as the POJO field name. | N/A |
| description                   | No         | A long description of the dimension.          | The dimension name |
| category                       | No         | Category for grouping                                  | N/A            |
| hidden                     | No          | Whether this dimension is exposed via API metadata | false |
| readAccess                    | No         | Read permission for the dimension.            | Allow All     |
| definition                    | Yes        | The definition of the dimension (See below)   | N/A           |
| type                          | No         | The data type of the field (See types)     | text          |
| tags                          | No        | An array of string based tags       | empty |

Dimension definitions are extensible objects that contain a _type_ field and one or more additional attributes.  Each type is tied to logic in Elide that generates one or more dimension columns in the table.  

Most dimension definitions can be represented with simple SQL expressions.  To simplify the common case, _definition_ can be mapped to a simple templated string instead of an object:

`definition: createdOn`

The SQL templates can reference columns:
1. by name: `column`
2. by a dot separated path to a column: `player.session.score`.

When using dot separated paths, all segments (excluding the final segment) represent _join_ names originating from the current table.  The final segment represents the column in the final joined table.

### Time Dimensions

Time dimensions are dimensions of type _time_.  Time dimensions can converted at query time to different grains by supplying a time definition object:

```
{
   name : createdOn
   definition : create_on
   type: time
   grains: [{{day}}, {{month}}, {{year}}]
}
```

Grains are themselves objects.  Grains can have SQL expressions that can substitute _column_ with the dimension definition expression:
```
{

   grain: month
   sql: "PARSEDATETIME(FORMATDATETIME(${column}, 'yyyy-MM-01'), 'yyyy-MM-dd')"
}
```

It is recommended to define grains once in _variables.hjson_.

## Metric and dimension types

| Type | Applies To |
| ---- | ----------- |
| time | Dimension  |
| text | Dimension  |
| coordinate | Dimension |
| number | Metric & Dimension |
| boolean | Dimension |

## Variable Substitution

The _variable.hjson_ file can be used to substitute repeated fragments of HJSON in model files.  The file format looks like:

```
{
   foo: [1, 2, 3]
   bar: blah
}
```

Model files are effectively [handlebars](https://handlebarsjs.com/) templates.  Model files can reference variable names in _variable.hjson_ using the syntax: `{{variableName}}`.

### Permission Rules

Elide permissions for tables and fields are limited to read permissions only.    Permission roles and rules are defined in the _security.hjson_ file:

```
{
    roles : [admin, guest, member]
    rules: [
        {
            type: filter //type is optional.
            filter: company_id=in=${principal.companies}
            name: User belongs to company
        },
        {
            filter: id==${principal.id}
            name: Principal is owner
        },
    ]
}
```

Roles are a list of roles that will map to security checks in Elide with:
1. Names that match "Principal belongs to role _ROLE_".
2. Logic that checks the Principal role membership.

Rules are a list of RSQL filter expression templates that support property expansion on the principal object.  These RSQL filter expressions are mapped to `FilterExpressionCheck` classes in Elide.

### Code Generation & Input Validation

Elide POJOs are generated using [handlebars](https://handlebarsjs.com/) templates.  The values taken from the configuration file are restricted depending on how they are used:
1. Minimal restrictions:
  - Quotations will be escaped
  - reserved words will be banned (to limit SQL fragments to DQL).
2. Names only
  - The text must be alphanumeric name starting with a letter only.

SQL expressions will be limited to DQL queries only (other reserved words will be prohibited).

### Configuration Tamper Proofing

Elide will support an option to verify the signature of model files using a public key in the server trust store.  The entire models directory can be signed by creating a tarball archive of the directory and signing it with PGP.  The _model.asc_ file will be included in the model directory (but excluded from the tar archive when producing the signature).  

## Validation

Elide will validate the following:

1. The HJSON files conform to valid JSON schemas according to their file type.
2. Model and field names references are valid in expressions.
3. Permissions expressions are defined.
5. Handlebars templates are fully resolved.
6. Serialization of HJSON to types is valid.
7. Enumeration values (where used) are valid.
8. Tags in the elide namespace are restricted to a small set of defined tags.

The logic which performs validation should be available both as a Java library and a Java command line executable.

## Caveats 
1. First version of config will not define entities.
2. Complex metrics like funnels and retention are not supported except through direct SQL configuration.
3. Parameterized metric definition is not defined.

### PR DESCRIPTION
**Pull request by AvaniMakwana**
Authored-by: AvaniMakwana

One of the multiple PRs that will resolve #1178

Description
- Add Table, Security & variable config validation against schema

Motivation and Context
Please refer #1178

How Has This Been Tested?
Unit tests have been included.

License
I confirm that this contribution is made under an Apache 2.0 license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: elide-contrib/elide-dynamic-config-helpers/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/DynamicConfigHelpers.java [hunk 1]
```diff
@@ -190,4 +228,45 @@ public static String readConfigFile(File configFile) {
...
+        }
+        return results.isSuccess();
+    }
+
+    private static JsonSchema loadSchema(String resource) {
+        ObjectMapper objectMapper = new ObjectMapper();
+        Reader reader = new InputStreamReader(DynamicConfigHelpers.class.getResourceAsStream(resource));
+        try {
+            return FACTORY.getJsonSchema(objectMapper.readTree(reader));
+        } catch (IOException e) {
```
Comment (moizarafat) [inline elide-contrib/elide-dynamic-config-helpers/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/DynamicConfigHelpers.java:263]: we should let these exceptions propagate and fail.

File: elide-contrib/elide-dynamic-config-helpers/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/DynamicConfigHelpers.java [hunk 2]
```diff
@@ -64,14 +89,19 @@ public static String formatFilePath(String basePath) {
...
+            throws JsonProcessingException, ProcessingException {
         String filePath = basePath + VARIABLE_CONFIG_PATH;
         File variableFile = new File(filePath);
         if (variableFile.exists()) {
             String jsonConfig = hjsonToJson(readConfigFile(variableFile));
-            return getModelPojo(jsonConfig, Map.class);
+            if (verifySchema(VARIABLE, jsonConfig)) {
+                return getModelPojo(jsonConfig, Map.class);
+            }
+            return null;
```
Comment (moizarafat) [inline elide-contrib/elide-dynamic-config-helpers/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/DynamicConfigHelpers.java:104]: instead of multiple return we can store results in a variable,  and then have single return.

File: elide-contrib/elide-dynamic-config-helpers/src/test/java/com/yahoo/elide/contrib/dynamicconfighelpers/validator/DynamicConfigValidatorTest.java [hunk 3]
```diff
@@ -107,7 +107,7 @@ public void testBadSecurityRoleConfig() {
 
     @Test
     public void testBadTableConfigJoinType() {
-        assertThrows(ValueInstantiationException.class, () -> DynamicConfigValidator
+        assertThrows(NullPointerException.class, () -> DynamicConfigValidator
```
Comment (moizarafat) [inline elide-contrib/elide-dynamic-config-helpers/src/test/java/com/yahoo/elide/contrib/dynamicconfighelpers/validator/DynamicConfigValidatorTest.java:110]: is this change needed?

File: elide-contrib/elide-dynamic-config-helpers/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/DynamicConfigHelpers.java [hunk 4]
```diff
@@ -190,4 +230,46 @@ public static String readConfigFile(File configFile) {
...
+    }
+
+    private static JsonSchema loadSchema(String resource) {
+        ObjectMapper objectMapper = new ObjectMapper();
+        Reader reader = new InputStreamReader(DynamicConfigHelpers.class.getResourceAsStream(resource));
+        try {
+            return FACTORY.getJsonSchema(objectMapper.readTree(reader));
+        } catch (IOException e) {
+            log.error("Error loading schema file " + resource + " to verify");
+            throw new RuntimeException(e.getMessage());
```
Comment (moizarafat) [inline elide-contrib/elide-dynamic-config-helpers/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/DynamicConfigHelpers.java:269]: we can throw IllegalStateException instead RunTime else codacy complains.

File: elide-contrib/elide-dynamic-config-helpers/src/test/java/com/yahoo/elide/contrib/dynamicconfighelpers/validator/DynamicConfigValidatorTest.java (hunk 3 above)
Comment (moizarafat) [reply elide-contrib/elide-dynamic-config-helpers/src/test/java/com/yahoo/elide/contrib/dynamicconfighelpers/validator/DynamicConfigValidatorTest.java:110]: i don't see any other related change in main class.

File: elide-contrib/elide-dynamic-config-helpers/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/DynamicConfigHelpers.java [hunk 5]
```diff
@@ -190,4 +230,46 @@ public static String readConfigFile(File configFile) {
...
+        } catch (ProcessingException e) {
+            log.error("Error loading schema file " + resource + " to verify");
+            throw new RuntimeException(e.getMessage());
```
Comment (moizarafat) [inline elide-contrib/elide-dynamic-config-helpers/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/DynamicConfigHelpers.java:272]: same as previous comment.

File: elide-contrib/elide-dynamic-config-helpers/src/test/java/com/yahoo/elide/contrib/dynamicconfighelpers/validator/DynamicConfigValidatorTest.java (hunk 3 above)
Comment (AvaniMakwana) [reply elide-contrib/elide-dynamic-config-helpers/src/test/java/com/yahoo/elide/contrib/dynamicconfighelpers/validator/DynamicConfigValidatorTest.java:110]: the function throws NPE. hence the change.

File: elide-contrib/elide-dynamic-config-helpers/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/validator/DynamicConfigValidator.java [hunk 6]
```diff
@@ -14,6 +14,7 @@
 import com.yahoo.elide.contrib.dynamicconfighelpers.model.Table;
 
 import com.fasterxml.jackson.core.JsonProcessingException;
+import com.github.fge.jsonschema.core.exceptions.ProcessingException;
```
Comment (aklish) [inline elide-contrib/elide-dynamic-config-helpers/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/validator/DynamicConfigValidator.java:17]: We probably don't want to leak the dependency on jsonschema outside the validation logic.

File: elide-contrib/elide-dynamic-config-helpers/src/test/java/com/yahoo/elide/contrib/dynamicconfighelpers/validator/DynamicConfigValidatorTest.java (hunk 3 above)
Comment (aklish) [reply elide-contrib/elide-dynamic-config-helpers/src/test/java/com/yahoo/elide/contrib/dynamicconfighelpers/validator/DynamicConfigValidatorTest.java:108]: It looks very odd for a test to _expect_ a NPE.  NPE should never happen in code - so this test looks wrong or something is off.

File: elide-contrib/elide-dynamic-config-helpers/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/DynamicConfigHelpers.java [hunk 7]
```diff
@@ -190,4 +230,46 @@ public static String readConfigFile(File configFile) {
...
+
+    private static String hjsonToJson(String hjson) {
+        return JsonValue.readHjson(hjson).toString();
+    }
+
+    private static <T> T getModelPojo(String jsonConfig, final Class<T> configPojo) throws JsonProcessingException {
+        return new ObjectMapper().readValue(jsonConfig, configPojo);
+    }
+
+    private static boolean verifySchema(String configType, String jsonConfig)
```
Comment (aklish) [inline elide-contrib/elide-dynamic-config-helpers/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/DynamicConfigHelpers.java:242]: Do we have tests for this new verify logic?

File: elide-contrib/elide-dynamic-config-helpers/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/DynamicConfigSchemaValidator.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+import com.github.fge.jsonschema.main.JsonSchemaFactory;
+
+import lombok.extern.slf4j.Slf4j;
+
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.Reader;
+
+@Slf4j
+public class DynamicConfigSchemaValidator {
```
Comment (aklish) [inline elide-contrib/elide-dynamic-config-helpers/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/DynamicConfigSchemaValidator.java:21]: Missing javadoc.

File: elide-contrib/elide-dynamic-config-helpers/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/DynamicConfigSchemaValidator.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+     * @return whether config is valid
+     * @throws IOException
+     * @throws ProcessingException
+     */
+    public static boolean verifySchema(String configType, String jsonConfig) throws IOException, ProcessingException {
+        ProcessingReport results = null;
+        boolean isSuccess = false;
+
+        switch (configType) {
+        case DynamicConfigHelpers.TABLE :
```
Comment (aklish) [inline elide-contrib/elide-dynamic-config-helpers/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/DynamicConfigSchemaValidator.java:51]: Instead of configType being a string, we should make it an enum.

File: elide-contrib/elide-dynamic-config-helpers/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/DynamicConfigSchemaValidator.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+    /**
+     *  Verify config against schema.
+     * @param configType
+     * @param jsonConfig
+     * @return whether config is valid
+     * @throws IOException
+     * @throws ProcessingException
+     */
+    public static boolean verifySchema(String configType, String jsonConfig) throws IOException, ProcessingException {
```
Comment (aklish) [inline elide-contrib/elide-dynamic-config-helpers/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/DynamicConfigSchemaValidator.java:46]: Is there a reason all of this has to be static?  It will be easier to mock if it were not.

File: elide-contrib/elide-dynamic-config-helpers/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/DynamicConfigHelpers.java [hunk 11]
```diff
@@ -36,6 +38,9 @@
     public static final String SECURITY_CONFIG_PATH = "security.hjson";
     public static final String VARIABLE_CONFIG_PATH = "variables.hjson";
     public static final String NEW_LINE = "\n";
+    public static final String TABLE = "table";
```
Comment (aklish) [inline elide-contrib/elide-dynamic-config-helpers/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/DynamicConfigHelpers.java:41]: Make these an enum.

File: elide-contrib/elide-dynamic-config-helpers/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/DynamicConfigHelpers.java [hunk 12]
```diff
@@ -121,8 +135,17 @@ public static ElideTableConfig getElideTablePojo(String basePath, Map<String, Ob
...
-        return getModelPojo(jsonConfig, ElideTableConfig.class);
+        try {
+            if (DynamicConfigSchemaValidator.verifySchema(TABLE, jsonConfig)) {
+                table = getModelPojo(jsonConfig, ElideTableConfig.class);
+            }
+        } catch (ProcessingException e) {
+            log.error("Error Validating Table config : " + e.getMessage());
+            throw new IOException(e);
+        }
+        return (table == null ? new ElideTableConfig() : table);
```
Comment (aklish) [inline elide-contrib/elide-dynamic-config-helpers/src/main/java/com/yahoo/elide/contrib/dynamicconfighelpers/DynamicConfigHelpers.java:148]: If you initialize table to new ElideTableConfig, you can remove this check for null at the end.

**Comment by moizarafat** [comment]
Closed in favor of https://github.com/yahoo/elide/pull/1449
